feat(ivy): support context discovery for containers & ICU expressions

ocombe opened the pull request:
## PR Type
What kind of change does this PR introduce?

- [x] Feature


## What is the current behavior?
Context discovery is only available on elements.

## What is the new behavior?
This PR adds support for containers and ICU expressions.

## Does this PR introduce a breaking change?
- [x] No

mary-poppins commented:
You can preview 536b3cf at https://pr27644-536b3cf.ngbuilds.io/.

mary-poppins commented:
You can preview 0d200e9 at https://pr27644-0d200e9.ngbuilds.io/.

pkozlowski-opensource commented on packages/core/test/render3/discovery_utils_spec.ts:
```diff
@@ -289,12 +308,32 @@ describe('discovery utils', () => {
+      expect(lContext).toBeDefined();
+      expect(lContext.native as any).toBe(templateComment);
+    });
+
```
Shouldn't we add a test for `<ng-container>` ?

mary-poppins commented:
You can preview 3bcc956 at https://pr27644-3bcc956.ngbuilds.io/.

mary-poppins commented:
You can preview 756b271 at https://pr27644-756b271.ngbuilds.io/.

mary-poppins commented:
You can preview f728408 at https://pr27644-f728408.ngbuilds.io/.

mary-poppins commented:
You can preview 1ab5f8b at https://pr27644-1ab5f8b.ngbuilds.io/.

pkozlowski-opensource commented on packages/core/src/render3/context_discovery.ts:
```diff
@@ -209,7 +207,13 @@ function traverseNextElement(tNode: TNode): TNode|null {
     return tNode.next;
   } else if (tNode.parent) {
-    return tNode.parent.next || null;
+    while (tNode) {
```
It is not clear to me why we've got this logic. Could you please a comment here (or, better yet, add an utility function with a good name and description)?

ocombe replied:
The issue happens when we close 2 nodes. Let's take the following template:
```html
<div>
  <span>text</span>
</div>
<component/>
```

The function will do that:
div --> child (span)
span --> child (text)
text --> child? no, next? no, parent (span).next --> no, return null
In this case, we need to find the next parent that has a "next" TNode if we want to have a chance to find the component.

kara replied:
Would the intent be clearer if we wrote it this way?

```ts
if (tNode.child) {
   return tNode.child;
} else if (tNode.next) {   
   return tNode.next;
} else {
   while (tNode.parent && !tNode.parent.next) {
       tNode = tNode.parent;
   }
   return tNode.parent && tNode.parent.next;
}
```

Agree with @pkozlowski-opensource that a comment might help here, so we don't forget why.

pkozlowski-opensource commented on packages/core/src/render3/instructions.ts:
```diff
@@ -212,24 +212,24 @@ export function createNodeAtIndex(
-        previousOrParentTNode.next = tNode;
-      }
+  // Now link ourselves into the tree.
+  const previousOrParentTNode = getPreviousOrParentTNode();
```
This logic here starts to be hard to decipher for me... How about extracting it to a separate function with a descriptive name and more comments?

ocombe replied:
I've just changed it so that we still link ourselves into the tree even if the tNode exists, otherwise we might end up with element that point to unexisting tNodes when we combine i18n (that changes the DOM at runtime) and reusing the same template multiple times. That's especially true when we use ICU expressions that will update the DOM during the update phase

kara replied:
^ Sounds like a good candidate for a comment

pkozlowski-opensource commented on packages/core/test/render3/discovery_utils_spec.ts:
```diff
@@ -289,12 +299,46 @@ describe('discovery utils', () => {
+
+    it('should work on templates', () => {
+      const templateComment = Array.from(fixture.hostElement.childNodes)
+                                  .find((node: ChildNode) => node.nodeType === Node.COMMENT_NODE);
```
not, feel free to ignore / something to consider: personally I much prefer small, focused tests where I've got both setup and asserts in one place. So myself I would create a mini-component in each test with just one element that we need to test. This would:
* allow us to have test setup and asserts in one place (so I don't have to scroll 200 LOC up to see what is in the template;
* you could remove nodes filtering as you could have just one node.

Again, just a personal preference.

ocombe replied:
true, but in this case it was nice to have a big mix up of components/i18n/containers, it helped me find and fix a few bugs (we should probably have a better integration test in the i18n tests)

mary-poppins commented:
You can preview de87765 at https://pr27644-de87765.ngbuilds.io/.

kara reviewed (changes requested):
I think there are 2 or 3 more fixmes for FW-665 that need to be removed in `element_spec`.

ocombe committed:
feat(ivy): support context discovery for containers & ICU expressions

Context discovery was only available on elements. This PR adds support for containers and ICU expressions.
FW-378 #resolve
FW-665 #comment linker integration tests

ocombe committed:
fixup! feat(ivy): support context discovery for containers & ICU expressions

mary-poppins commented:
You can preview 6ed543a at https://pr27644-6ed543a.ngbuilds.io/.

kara reviewed (approved):
LGTM

mary-poppins commented:
You can preview 6c7a696 at https://pr27644-6c7a696.ngbuilds.io/.

kara commented:
[presubmit](http://test/OCL:225867300:BASE:225867454:1545077539213:12588d37)

JoostK commented on packages/core/src/render3/context_discovery.ts:
```diff
@@ -208,10 +206,15 @@ function traverseNextElement(tNode: TNode): TNode|null {
+    while (tNode.parent && !tNode.parent.next) {
+      tNode = tNode.parent;
+    }
+    return tNode.parent && tNode.parent.next;
```
This should be equivalent and avoid repeated `parent` reads.

```typescript
      tNode = tNode.parent;
      while (tNode && !tNode.next) {
        tNode = tNode.parent;
      }
      return tNode && tNode.next;
```
